### PR TITLE
Contest API Rust, UI

### PR DESCRIPTION
**Pull request by Valokoodari**

### CODE REVIEW

File: api/schemas.yaml [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+      description: >
+        Status of the submission, e.g. "PENDING" or "READY".
+    pending:
+      type: boolean
+      description: >
+        Whether the submission is still being graded,
+        and the information may change soon.
+    test_progress:
+      description: >
+        > An optional property describing the progress of the testing phase.
```
Comment (H4m5t3r) [inline api/schemas.yaml:202]: There's an extra '>' here

Comment (ollpu) [reply api/schemas.yaml:202]: This comes from @kluopaja's PR

File: src/ui/courses.rs [hunk 2]
```diff
@@ -72,13 +72,27 @@ pub fn create_course_item_table(list: &[CourseItemRaw]) -> Result<Table> {
...
+                    }
+                    _ => {
+                        table.add_row(vec![
+                            TableCell::from(id).align(TableAlign::Right),
+                            TableCell::from(name),
+                            TableCell::styled(styled_task_status(status)),
+                            TableCell::from(link),
+                        ]);
+                    }
+                }
```
Comment (ollpu) [inline src/ui/courses.rs:95]: Could this be done by hiding columns? We can add helper functions to `TableCell` to make that easier with `styled`

**Review by kluopaja — APPROVED** [review]

File: src/ui/courses.rs [hunk 3]
```diff
@@ -25,8 +28,8 @@ pub fn list_courses(ui: &mut Ui<impl RP>) -> Result<()> {
     Ok(())
 }
 
-pub fn list_course_content(ui: &mut Ui<impl RP>, course_id: &str) -> Result<()> {
-    let course = service::course_content(&mut ui.res, course_id)?;
+pub fn list_content(ui: &mut Ui<impl RP>, scope: &Scope) -> Result<()> {
+    let course = service::course_content(&mut ui.res, scope)?;
```
[the remark this answers is not in the document]
Comment (ollpu) [reply src/ui/courses.rs:32]: It is also possible to specialize the messages based on scope:

```rust
return Ok(writeln!(ui.term, match scope {
    Scope::Course(_) => "No course content!",
    Scope::Contest(_) => "No contest content!",
})?);
```
Just "No content!" works quite well here, though.

**Review by kluopaja — COMMENTED** [review]
In `src/entities/course.rs` the struct should be renamed e.g. `CourseContent` -> `ScopeContent` as they can now also hold contest data. Also the path should be renamed.

File: src/ui/util.rs [hunk 4]
```diff
@@ -14,3 +16,37 @@ pub fn format_code_time(time: Option<u64>) -> String {
...
+            points if points > 10 => style(points.to_string()).yellow(),
+            _ => style(points.to_string()).red(),
+        };
+    }
+
+    match status {
+        Some(TaskStatus::Pass) => style("+".to_string()).bold().green(),
+        Some(TaskStatus::Fail) => style("X".to_string()).bold().red(),
+        Some(TaskStatus::None) => style("-".to_string()).bold().dim(),
+        None => style("-".to_string()).dim(),
```
Comment (ollpu) [inline src/ui/util.rs:50]: Intentional for the `None` case to be different?
```suggestion
        Some(TaskStatus::None) | None => style("-".to_string()).bold().dim(),
```

Comment (ollpu) [reply src/ui/util.rs:50]: We also discussed making points bold

**Review by anttoh — APPROVED** [review]

**Review by H4m5t3r — APPROVED** [review]

**Pull request merged**